src/device_trezor/device_trezor.cpp

```diff
@@ -0,0 +1,354 @@
+
+    std::shared_ptr<messages::monero::MoneroAddress> device_trezor::get_address(
+        boost::optional<std::vector<uint32_t>> path,
+        boost::optional<cryptonote::network_type> network_type){
+      AUTO_LOCK_CMD();
+      require_connected();
+      test_ping();
+
+      auto req = std::make_shared<messages::monero::MoneroGetAddress>();
+      this->set_msg_addr<messages::monero::MoneroGetAddress>(req.get(), path, network_type);
+
+      auto response = this->client_exchange<messages::monero::MoneroAddress>(req);
+      MDEBUG("Get address response received");
+      return response;
+    }
+
+    std::shared_ptr<messages::monero::MoneroWatchKey> device_trezor::get_watch_key(
+        boost::optional<std::vector<uint32_t>> path,
+        boost::optional<cryptonote::network_type> network_type){
+      AUTO_LOCK_CMD();
+      require_connected();
+      test_ping();
+
+      auto req = std::make_shared<messages::monero::MoneroGetWatchKey>();
+      this->set_msg_addr<messages::monero::MoneroGetWatchKey>(req.get(), path, network_type);
+
+      auto response = this->client_exchange<messages::monero::MoneroWatchKey>(req);
+      MDEBUG("Get watch key response received");
+      return response;
+    }
+
+    void device_trezor::ki_sync(wallet_shim * wallet,
+                                const std::vector<tools::wallet2::transfer_details> & transfers,
+                                hw::device_cold::exported_key_image & ski)
+    {
+      AUTO_LOCK_CMD();
+      require_connected();
+      test_ping();
+
+      std::shared_ptr<messages::monero::MoneroKeyImageExportInitRequest> req;
+
+      std::vector<protocol::ki::MoneroTransferDetails> mtds;
+      std::vector<protocol::ki::MoneroExportedKeyImage> kis;
+      protocol::ki::key_image_data(wallet, transfers, mtds);
+
+      bool res = protocol::ki::generate_commitment(mtds, transfers, req);
+      this->set_msg_addr<messages::monero::MoneroKeyImageExportInitRequest>(req.get());
+      auto ack1 = this->client_exchange<messages::monero::MoneroKeyImageExportInitAck>(req);
+
+      const auto batch_size = 10;
+      const auto num_batches = static_cast<uint64_t>(ceil(mtds.size() / static_cast<double>(batch_size)));
+      for(uint64_t cur = 0; cur < num_batches; ++cur){
+        auto step_req = std::make_shared<messages::monero::MoneroKeyImageSyncStepRequest>();
+        auto idx_finish = std::min(static_cast<uint64_t>((cur + 1) * batch_size), static_cast<uint64_t>(mtds.size()));
+        for(uint64_t idx = cur * batch_size; idx < idx_finish; ++idx){
+          auto added_tdis = step_req->add_tdis();
+          *added_tdis = mtds[idx];
+        }
+
+        auto step_ack = this->client_exchange<messages::monero::MoneroKeyImageSyncStepAck>(step_req);
+        auto kis_size = step_ack->kis_size();
+        for(int i = 0; i < kis_size; ++i){
+          auto ckis = step_ack->kis(i);
+          kis.push_back(ckis);
```

Comment:
reserve could be called since the size of known in advance.

Reply:
done